### PR TITLE
[distributions] Fix scalar bugs in torch.distributions.transforms etc.

### PR DESCRIPTION
**Pull request by fritzo**
This attempts to fix a number of edge-case bugs in `Transforms` and `ConstraintRegistry`:

1. ~~Avoids `broadcast_all()` on statically defined transforms, since these must work with all tensor types (float/double, cpu/cuda), and often are defined only via integers, e.g. `AffineTransform(0, -1)`. This PR aims to support `numbers.Number` as `.loc` and `.scale` of `AffineTransform`.~~
  EDIT This will be addressed in a later PR.

2. Support scalars in the `.log_bas_det_jacobian()` method of `identity_transform = ComposeTransform([])`. This had not been updated since PyTorch supported scalars.

This also updates `test_distributions.py` to use more modern PyTorch syntax, and fixes some broken documentation. 

## Tested

- added a few `Transform`s and `TransformedDistribution`s with parameters that are `numbers.Number`

### CODE REVIEW

**Comment by fritzo** [comment]
cc @alicanb @apaszke I've attempted to use more modern idioms here, but I could use help writing more thorough tests if you have any ideas.

File: test/test_distributions.py [hunk 1]
```diff
@@ -319,6 +319,11 @@ def is_all_nan(tensor):
             'transforms': [AffineTransform(variable(torch.randn(3, 5)), variable(torch.randn(3, 5))),
                            ExpTransform()],
         },
+        {
+            'base_distribution': Normal(variable(torch.randn(2, 3, 5), requires_grad=True),
+                                        variable(torch.randn(2, 3, 5).abs(), requires_grad=True)),
```
Comment (apaszke) [inline test/test_distributions.py]: You can remove `variable` and put `requires_grad` inside `randn`

File: test/test_distributions.py [hunk 2]
```diff
@@ -3263,6 +3301,10 @@ def test_transform_to(self):
             y2 = t(x2)
             self.assertEqual(y, y2, message="Error in transform_to({}) pseudoinverse".format(constraint))
 
+            if t.bijective:
+                j = t.log_abs_det_jacobian(x, y)
+                self.assertEqual(j.shape, x.shape[:x.dim() - t.event_dim])
```
Comment (apaszke) [inline test/test_distributions.py]: Isn’t this redundant with `test_biject_to`?

File: torch/distributions/transforms.py [hunk 3]
```diff
@@ -354,19 +355,34 @@ class AffineTransform(Transform):
...
+        else:
+            if (self.scale != other.scale).any().item():
+                return False
+
+        return True
 
     @property
     def sign(self):
+        if isinstance(self.scale, numbers.Number):
+            return (self.scale > 0) - (self.scale < 0)
```
Comment (apaszke) [inline torch/distributions/transforms.py]: Uh please don’t do arithmetic on Boolean values.

File: torch/distributions/transforms.py [hunk 4]
```diff
@@ -376,8 +392,11 @@ def _inverse(self, y):
         return (y - self.loc) / self.scale
 
     def log_abs_det_jacobian(self, x, y):
-        result = torch.abs(self.scale).log()
         shape = x.shape
+        scale = self.scale
+        if isinstance(scale, numbers.Number):
+            scale = x.new_empty(shape).fill_(scale)
+        result = torch.abs(scale).log()
```
Comment (apaszke) [inline torch/distributions/transforms.py]: Minor optimization: abs can be lifted to the scalar case and would be cheaper

File: torch/distributions/transforms.py [hunk 5]
```diff
@@ -354,19 +355,34 @@ class AffineTransform(Transform):
 
     def __init__(self, loc, scale, event_dim=0, cache_size=0):
         super(AffineTransform, self).__init__(cache_size=cache_size)
-        self.loc, self.scale = broadcast_all(loc, scale)
+        self.loc = loc
+        self.scale = scale
```
Comment (apaszke) [inline torch/distributions/transforms.py]: Why can’t we promote them to tensors here? Not having clearly typed attributes will be a constant source of bugs

File: torch/distributions/transforms.py (hunk 3 above)
Comment (fritzo) [reply torch/distributions/transforms.py]: Ok, I'll switch to a conditional, but this is the recommended way of writing `cmp(self.scale, 0)` in a Python 2/3 compatible way: https://docs.python.org/3.0/whatsnew/3.0.html#ordering-comparisons

File: torch/distributions/transforms.py (hunk 4 above)
Comment (fritzo) [reply torch/distributions/transforms.py]: Good point, I'll move the `log` in there too.

File: torch/distributions/transforms.py (hunk 5 above)
Comment (apaszke) [reply torch/distributions/transforms.py]: I don't think you need to know that exactly. If they are tensors, then keep them like it, and if they are scalars then we are actually in the process of adding support for implicit casting, so you can probably safely use the default tensor type (`torch.tensor(loc)`). See #5864.

Comment (fritzo) [reply torch/distributions/transforms.py]: Oh great, are you're planning to add casting from e.g. `torch.FloatTensor` to `torch.cuda.DoubleTensor` before 0.4 release? If so, then I'll revert this 3-line change.

Comment (fritzo) [reply torch/distributions/transforms.py]: After discussing offline I've reverted this change. I'll move this to a separate PR for further discussion.

**Review by apaszke — APPROVED** [review]

**Pull request merged**